lumigator/python/mzai/jobs/inference/inference.py

```diff
@@ -0,0 +1,161 @@
+"""python job to run batch inference"""
+
+import argparse
+import functools
+import json
+import os
+import time
+from collections.abc import Iterable
+from pathlib import Path
+
+import s3fs
+from box import Box
+from datasets import load_from_disk
+from loguru import logger
+from model_clients import (
+    BaseModelClient,
+    MistralModelClient,
+    OpenAIModelClient,
+)
+
+# SummarizationPipelineModelClient,
+# HuggingFaceModelClient,
+from tqdm import tqdm
+
+
+def timer(func):
+    """Decorator which times the execution of the wrapped func.
+    Execution time is logged and also returned together with func's returned value
+    (output will be a tuple).
+    """
+
+    @functools.wraps(func)
+    def wrapper_timer(*args, **kwargs):
+        tic = time.perf_counter()
+        value = func(*args, **kwargs)
+        toc = time.perf_counter()
+        elapsed_time = toc - tic
+        logger.info(f"Elapsed time for {func.__name__}: {elapsed_time:0.4f} seconds")
+        return value, elapsed_time
+
+    return wrapper_timer
+
+
+@timer
+def predict(dataset_iterable: Iterable, model_client: BaseModelClient) -> list:
+    predictions = []
+
+    for sample_txt in dataset_iterable:
+        predictions.append(model_client.predict(sample_txt))
+
+    return predictions
+
+
+def save_outputs(config: Box, inference_results: dict) -> Path:
+    storage_path = config.evaluation.storage_path
+
+    def save_to_disk(local_path: Path):
+        logger.info(f"Storing into {local_path}...")
+        local_path.parent.mkdir(exist_ok=True, parents=True)
+        with local_path.open("w") as f:
+            json.dump(inference_results, f)
+
+    def save_to_s3(local_path: Path, storage_path: str):
+        s3 = s3fs.S3FileSystem()
+        if storage_path.endswith("/"):
+            storage_path = "s3://" + str(
+                Path(storage_path[5:]) / config.name / "inference_results.json"
+            )
+        logger.info(f"Storing into {storage_path}...")
+        s3.put_file(local_path, storage_path)
+
+    # generate local temp file ANYWAY
+    # (we don't want to lose all eval data if there is an issue wth s3)
+    local_path = Path(
```

Comment:
I think this is mostly an artifact from the old lm-buddy use-case: we used temp files in the past but in this specific case I wanted to have an actual file which is deleted after the upload, so that if anything breaks on the S3 side the file is at least available locally.
However, this is run by a worker now and even if the file is stored locally it will be deleted when the job ends... So on the one hand we don't have to worry about that, on the other one though we'd like to make sure the file remains available for the above reasons?

Reply:
Yup, the file will still be available, mkdtemp just offers a cleaner way to get to it

Reply:
Cool, thanks! I'll add this into the next PR then!